waiter/src/waiter/process_request.clj

```diff
@@ -483,6 +482,57 @@
             (meters/mark! (metrics/waiter-meter "core" "process-errors"))
             (utils/exception->response e request)))))))
 
+(defn service-id->interstitial-promise
+  "TODO (shams) docstring"
+  [service-id->interstitial-promise-atom service-id]
+  (get @service-id->interstitial-promise-atom service-id))
+
+(defn start-service-interstitial!
+  "TODO (shams) docstring"
+  [service-id->interstitial-promise-atom service-id interstitial-secs]
+  (if-let [existing-interstitial-timed-out-promise
+           (service-id->interstitial-promise service-id->interstitial-promise-atom service-id)]
+    existing-interstitial-timed-out-promise
+    (let [interstitial-timed-out-promise (promise)]
+      (->> (fn [service-id->interstitial-promise]
+             (log/info "creating interstitial promise for" service-id)
+             (if-not (contains? service-id->interstitial-promise service-id)
+               ;; TODO (shams) trigger call to start service in follow-up PR, so we can move it out of prepare instance
+               (assoc service-id->interstitial-promise service-id interstitial-timed-out-promise)
+               service-id->interstitial-promise))
+           (swap! service-id->interstitial-promise-atom))
+      ;; TODO (shams) trigger deliver as soon as we have a healthy instance for this service,
+      ;; TODO (shams) e.g. by running a separate go-block that listens to scheduler messages
+      (if (pos? interstitial-secs)
+        (async/go
+          (-> interstitial-secs t/seconds t/in-millis async/timeout async/<!)
```

Comment:
Did we decide that `interstitial-secs` is based upon the time we started the service?  Do we need to account for that here?

Reply:
Eventually this block will also include starting the service, so it will include the starting time as marathon delay + time to actually start service.

Reply:
I decided to not implement starting the service in this PR. Our current logic relies on an instance requesting for an instance to decide on starting a service, that logic is not easy to port when we eagerly respond with an interstitial but expect an outstanding service to trigger starting an instance.